index.bs

```diff
@@ -795,6 +743,562 @@ the following list:
+ messages being handled until the backpressure is released. After an implementation-defined amount
+ of data is buffered, the client will stop reading from the underlying network connection resulting
+ in network-layer backpressure being applied.
+</div>
+
+<hr>
+
+When [=the WebSocket closing handshake is started=] for a {{WebSocketStream}} |stream|, the user
+agent must [=queue a task=] to change the |stream|'s [=WebSocketStream/ready state=] to
+{{WebSocket/CLOSING}} if it is not already. [[!WSP]]
+
+<hr>
+
+<div algorithm="handle WebSocketStream connection close">
+
+When [=the WebSocket connection is closed=] for a {{WebSocketStream}} |stream|, possibly
+ [=cleanly=], the user agent must [=queue a task=] to run the following substeps:
+
+ 1. Change the [=WebSocketStream/ready state=] to {{WebSocket/CLOSED}} (3).
+ 1. If |stream|'s [=WebSocketStream/handshake aborted flag=] is set, then return.
+ 1. If |stream|'s [=WebSocketStream/was ever connected flag=] is not set, then [=reject=] |stream|'s
+    [=opened promise=] with a [=new=] WebSocketError.
+ 1. Let |code| be [=the WebSocket connection close code=].
+ 1. Let |reason| be the result of applying [=UTF-8 decode without BOM=] to [=the WebSocket
+    connection close reason=].
+ 1. If the connection was closed [=cleanly=],
+  1. [=ReadableStream/Close=] |stream|'s [=WebSocketStream/readable stream=].
+  1. [=WritableStream/Error=] |stream|'s [=WebSocketStream/writable stream=] with an
+     "{{InvalidStateError}}" {{DOMException}} indicating that a closed {{WebSocketStream}} cannot be
+     written to.
+  1. [=Resolve=] |stream|'s [=closed promise=] with a [=new=] {{WebSocketCloseInfo}} constructed
+     with |code| and |reason|.
+ 1. Otherwise,
+  1. [=ReadableStream/Error=] |stream|'s [=WebSocketStream/readable stream=] with a [=new=]
+     {{WebSocketError}} constructed with |code| and |reason|.
+  1. [=WritableStream/Error=] |stream|'s [=WebSocketStream/writable stream=] with a [=new=]
+     {{WebSocketError}} constructed with |code| and |reason|.
+  1. [=Reject=] |stream|'s [=closed promise=] with a [=new=] {{WebSocketError}} constructed with
+     |code| and |reason|.
+</div>
+
+<hr>
+
+The [=task source=] for all [=tasks=] <a lt="queue a task">queued</a> in this section is the
+[=WebSocket task source=].
+
+<hr>
+
+
+## Stream operations ## {#stream-operations}
+
+<div algorithm="pull algorithm">
+
+To <dfn>pull bytes</dfn> from a {{WebSocketStream}} |stream|, if |stream| is currently applying
+backpressure, release backpressure.
+
+ Note: If any messages are queued, one will be handled immediately as a result.
+</div>
+
+<hr>
+
+<div algorithm="cancel algorithm">
+
+To <dfn>cancel</dfn> a {{WebSocketStream}} |stream| given |reason|, [=close using reason=] giving
+|stream| and |reason|.
+</div>
+
+<hr>
+
+<div algorithm="write algorithm">
+
+To <dfn>write</dfn> |chunk| to a {{WebSocketStream}} |stream|, the user agent must run these steps:
+ 1. Let |promise| be a new promise.
```

Comment:
```suggestion
 1. Let |promise| be [=a new promise=] created in ....
```

where you need to fill in a realm.

Reply:
Done, although since the promise is purely internal, I'm not sure it makes a difference what realm it is created in.